design/design-sketch-oam.md

```diff
@@ -0,0 +1,334 @@
+      httpGet:
+        port: 8080
+        path: /healthz
+---
+apiVersion: core.oam.dev/v1alpha1
+kind: ManualScalerTrait
+metadata:
+  namespace: default
+  name: coolest-app-n3v3y
+spec:
+  replicaCount: 3
+---
+apiVersion: cache.crossplane.io/v1alpha1
+kind: RedisCluster
+metadata:
+  namespace: default
+  name: coolest-app-jnkf9
+spec:
+  classSelector:
+    matchLabels:
+      environment: production
+      region: europe
+  writeConnectionSecretToRef:
+    name: coolest-app-redis
+  engineVersion: "3.2
+```
+
+### Crossplane as an OAM Runtime
+
+This document proposes Crossplane initially support two kinds of component:
+`ContainerizedWorkload`, and Crossplane resource claims. Each containerized
+workload would be rendered to a series of Kubernetes resources (e.g. `Service`,
+`Deployment`, `Ingress`, etc). These resources would be created either in the
+Kubernetes control plane upon which Crossplane is running (similar to Rudr), or
+scheduled to an external Kubernetes cluster of which Crossplane is aware. The
+proposed `SchedulingScope` would determine where the resources were scheduled.
+Each resources that should be created on an external Kubernetes cluster would be
+represented in the Crossplane API server as a `KubernetesApplicationResource` as
+it is today. Note that these `KubernetesApplicationResource` resources would be
+owned (in the controller reference sense) by the `ContainerizedWorkload`; there
+would be no intermediate `KubernetesApplication`. `KubernetesApplication` would
+continue to exist as a distinct, lower level concept in Crossplane, as it is
+useful for scheduling and running arbitrary Kubernetes resources on an external
+cluster without the separation of concerns focused abstraction of an
+`ApplicationConfiguration`.
+
+## Open Questions
+
+The following questions are unresolved by this proposal:
+
+* Should Crossplane resource claims natively support the concept of OAM traits?
+  This proposal implies a distinct Kubernetes controller would reconcile any
+  `ContainerizedWorkload` resources, which would be required to maintain object
+  references to any configured traits. The proposal also implies that resource
+  claims _become_ a kind of extended workload (rather than being produced by a
+  kind of extended workload).
+* Should Crossplane resource claims natively support the concept of OAM scopes?
+  Similar to the above concerns, if an `ApplicationConfiguration` templates and
+  produces resource claims verbatim (as opposed to producing a resource claim
+  modified based on the supplied scopes), the controllers for said claims would
+  need to be aware of scopes in order to consider them when acting on the
+  resource claim.
+* What is the best approach to validate the array of components and traits under
+  the `ApplicationConfiguration` `components` field? One option may be to use a
+  validating webhook that finds the CRD that defines the templated kind and uses
+  its OpenAPI schema to validate it.
```

Comment:
Why don't we just use `WorkloadType` and `Trait` to define the schema? We can also use `validating webhook` to check the schema. In fact, as we use `ApplicationConfiguration`, we already have a layer between user application and K8s CRD. So `WorkloadType` and `Trait` didn't bring any burden, they can help unify different CRDs.